### PR TITLE
Make errors more fine-grained

### LINKED ISSUE #72: Errors hold strings
All errors hold a string as their contents. Better hold an enum to make it more semantic and to avoid unnecessary memory allocation and string copying, or at least use a `'static str`.
https://github.com/flavray/avro-rs/blob/7b820c13a1f25c7fd35850a6fed921aabc3ed103/src/de.rs#L14
https://github.com/flavray/avro-rs/blob/7b820c13a1f25c7fd35850a6fed921aabc3ed103/src/ser.rs#L29
https://github.com/flavray/avro-rs/blob/7b820c13a1f25c7fd35850a6fed921aabc3ed103/src/schema.rs#L17
https://github.com/flavray/avro-rs/blob/7b820c13a1f25c7fd35850a6fed921aabc3ed103/src/util.rs#L18
https://github.com/flavray/avro-rs/blob/7b820c13a1f25c7fd35850a6fed921aabc3ed103/src/util.rs#L32

### PR DESCRIPTION
**Pull request by cpcloud**
This PR makes a large variety of error variants.

This approach removes most allocations in error construction.

Closes #72.

### CODE REVIEW

**Comment by poros** [comment]
Wow, this is... a ton of work! I'll need a bit of time to go through everything, but thanks a lot in advance :)

Ah, I think this conflicted with your other PR.

File: tests/schema.rs [hunk 1]
```diff
@@ -759,7 +759,8 @@ fn test_root_error_is_not_swallowed_on_parse_error() -> Result<(), String> {
     let raw_schema = r#"/not/a/real/file"#;
     let error = Schema::parse_str(raw_schema).unwrap_err();
 
-    if let Error::JSON(e) = error {
+    // TODO: (#82) this should be a ParseSchemaError wrapping the JSON error
```
Comment (poros) [inline tests/schema.rs:762]: I don't understand this TODO. In addition, #82  has been closed already.

**Comment by poros** [comment]
HI @cpcloud , I'd really like to merge this PR and release the new version. I can update the changelog and the migration guide myself if you don't have time, but could you please answer my comment about that TODO?

**Comment by cpcloud** [comment]
Oh yeah! I should be able to get to this tomorrow morning!

File: tests/schema.rs (hunk 1 above)
Comment (cpcloud) [reply tests/schema.rs:762]: Might be leftover from a rebase.

Comment (cpcloud) [reply tests/schema.rs:762]: Removed.

**Comment by cpcloud** [comment]
@poros I'm not entirely sure how you want the new example to look. Shall I add a migration from, say `Error::Decode` to `Error::ReadDuration` or something like that?

**Comment by poros** [comment]
> @poros I'm not entirely sure how you want the new example to look. Shall I add a migration from, say `Error::Decode` to `Error::ReadDuration` or something like that?

Oh, no strong opinion, thanks for asking. It's just to help people upgrading to the new version. What you proposed works fine :)

**Comment by cpcloud** [comment]
@poros Updated the migration guide, let me know if it suffices.

**Comment by poros** [comment]
That works, thanks. I'll give the whole manual another read anyway. It's the first time I add one for this crate.

**Pull request merged**